pilot/pkg/xds/extension_discovery.go

```diff
@@ -0,0 +1,131 @@
+// limitations under the License.
+
+package xds
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	extensionservice "github.com/envoyproxy/go-control-plane/envoy/service/extension/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/cache/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/server/v3"
+	"google.golang.org/grpc/codes"
+	status "google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+)
+
+const (
+	// ExtensionConfigTypeURL for extension configs.
+	ExtensionConfigTypeURL = "type.googleapis.com/envoy.config.core.v3.TypedExtensionConfig"
+)
+
+// ExtensionServer is the main server instance.
+type ExtensionServer struct {
+	sync.Mutex
+	server.Server
+	server.CallbackFuncs
+	cache   *cache.LinearCache
+	configs map[string]*core.TypedExtensionConfig
+}
+
+var _ extensionservice.ExtensionConfigDiscoveryServiceServer = &ExtensionServer{}
+
+func NewExtensionServer() *ExtensionServer {
+	out := &ExtensionServer{}
+	out.cache = cache.NewLinearCache(ExtensionConfigTypeURL, cache.WithVersionPrefix(time.Now().Format(time.RFC3339)+"/"))
+	out.Server = server.NewServer(context.Background(), out.cache, out)
+	out.configs = make(map[string]*core.TypedExtensionConfig)
+	return out
+}
+
+func (es *ExtensionServer) StreamExtensionConfigs(stream extensionservice.ExtensionConfigDiscoveryService_StreamExtensionConfigsServer) error {
+	return es.Server.StreamHandler(stream, ExtensionConfigTypeURL)
+}
+func (es *ExtensionServer) DeltaExtensionConfigs(_ extensionservice.ExtensionConfigDiscoveryService_DeltaExtensionConfigsServer) error {
+	return status.Errorf(codes.Unimplemented, "not implemented")
+}
+func (es *ExtensionServer) FetchExtensionConfigs(ctx context.Context, req *discovery.DiscoveryRequest) (*discovery.DiscoveryResponse, error) {
+	req.TypeUrl = ExtensionConfigTypeURL
+	return es.Server.Fetch(ctx, req)
+}
+
+// OnStreamRequest qualifies the requested resource name by the namespace
+func (es *ExtensionServer) OnStreamRequest(streamID int64, req *discovery.DiscoveryRequest) error {
+	ns := ""
+	if req.Node.Metadata != nil {
+		if field, ok := req.Node.Metadata.Fields["NAMESPACE"]; ok {
+			ns = field.GetStringValue()
+		}
+	}
+	names := make([]string, 0, len(req.ResourceNames))
+	for _, name := range req.ResourceNames {
+		names = append(names, ns+"/"+name)
+	}
+	req.ResourceNames = names
+	return nil
+}
+
+// Update scans through all EnvoyFilters, identifies ECDS, and reconciles the serving cache.
+// This can be invoked from multiple go-routines executing AdsPushAll.
+func (es *ExtensionServer) Update(req *model.PushRequest) {
+	es.Lock()
+	defer es.Unlock()
+
+	staged := make(map[string]*core.TypedExtensionConfig)
+	allEnvoyFilters := req.Push.AllEnvoyFilters()
```

Comment:
why do you need all envoy filters for all namespaces? the logic you have below is also incorrect as it ignores workload labels.

Reply:
We need to list all ECDS configs, this is the way to do it without quadratic lookups. The logic is fine, since ECDS is on-demand, so the application in the second EnvoyFilter is sufficient to control the workload labels.

Reply:
under the assumption that people wont use workload label specific extensions?

Reply:
You cannot use extensions if you don't have the second application EnvoyFilter. The extra control would be for visibility of the binary itself. This code respects the namespace visibility, but not workload labels. I'm just not sure about the value of restricting the visibility of a binary in the same namespace. Even if you can see it, you can't apply it.